scripts/ws-hoard-upgrade.sh

```
@@ -1,49 +1,221 @@
+_ws_hoard_rollback() {
+    local hoard_dir="$1"
+    local backups_dir="$hoard_dir/.upgrade-backup"
+    [[ -d "$backups_dir" ]] || return 1
+    local latest
+    latest="$(find "$backups_dir" -mindepth 1 -maxdepth 1 -type d | sort | tail -n 1)"
+    [[ -n "$latest" ]] || return 1
+    local entry base
+    for entry in "$latest"/* "$latest"/.[!.]*; do
+        [[ -e "$entry" ]] || continue
+        base="$(basename "$entry")"
+        rm -rf "$hoard_dir/$base"
```

Shellcheck (0.11.0)

[warning] Use "${var:?}" to ensure this never expands to / .

(SC2115)